x-pack/filebeat/input/gcs/state.go

```diff
@@ -0,0 +1,94 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcs
+
+import (
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	maxFailedJobRetries int = 3
+)
+
+// state contains the the current state of the operation
+type state struct {
+	// Mutex lock to help in concurrent R/W
+	mu sync.Mutex
+	cp *Checkpoint
+}
+
+// Gcs sdks do not return results based on timestamps , but only based on lexicographic order
+// This forces us to maintain 2 different variables to calculate the exact checkpoint based on various scenarios
+type Checkpoint struct {
+	// name of the latest blob in alphabetical order
+	ObjectName string
+	// timestamp to denote which is the latest blob
+	LatestEntryTime *time.Time
+	// list of failed jobs due to unexpected errors/download errors
+	FailedJobs map[string]interface{}
+}
+
+func newState() *state {
+	return &state{
+		cp: &Checkpoint{
+			FailedJobs: make(map[string]interface{}),
+		},
+	}
+}
+
+// save, saves/updates the current state for cursor checkpoint
+func (s *state) save(name string, lastModifiedOn *time.Time) {
+	s.mu.Lock()
+	if s.cp.FailedJobs[name] == nil {
+		if len(s.cp.ObjectName) == 0 {
+			s.cp.ObjectName = name
+		} else if strings.ToLower(name) > strings.ToLower(s.cp.ObjectName) {
+			s.cp.ObjectName = name
+		}
+
+		if s.cp.LatestEntryTime == nil {
+			s.cp.LatestEntryTime = lastModifiedOn
+		} else if lastModifiedOn.After(*s.cp.LatestEntryTime) {
+			s.cp.LatestEntryTime = lastModifiedOn
+		}
+	} else {
+		// clear entry if this is a failed job
+		delete(s.cp.FailedJobs, name)
+	}
+	s.mu.Unlock()
+}
+
+// updateFailedJobs, adds a job name to a failedJobs map, which helps
+// in keeping track of failed jobs during edge cases when the state might
+// move ahead in timestamp & objectName due to successful operations from other workers.
+// A failed job will be re-tried a maximum of 3 times after which the
+// entry is removed from the map
+func (s *state) updateFailedJobs(jobName string) {
+	s.mu.Lock()
+	if s.cp.FailedJobs[jobName] == nil {
+		s.cp.FailedJobs[jobName] = 0
+	} else {
+		count := s.cp.FailedJobs[jobName].(int)
+		count++
+		if count > maxFailedJobRetries {
+			delete(s.cp.FailedJobs, jobName)
+		} else {
+			s.cp.FailedJobs[jobName] = count
+		}
+	}
```

Comment:
If `FailedJobs` can be `map[string]int`
```suggestion
	s.cp.FailedJobs[jobName]++
	if s.cp.FailedJobs[jobName] > maxFailedJobRetries {
		delete(s.cp.FailedJobs, jobName)
	}
```
What does it mean when the job has been removed from the map after it hits the max retry? Doesn't this mean that it gets reset to no failures?

Reply:
yes exactly, this is only for retrying failed jobs at a later time, if the job failed due to some unexpected error and not an internal api timeout. Internal timeout errors are handled by the sdk. This is just a fallback mechanism if there was an unexpected error. The fall back retry will occur a max of 3 time after which the job will be removed. We will have logs to catch all the errors.

Reply:
Thanks, that was what I was interpreting it as. I'm just wondering how an unfailed job is distinguished from a job that has failed more that `maxFailedJobRetries`. If the failure state is only kept in `s.cp.FailedJobs` I don't see how it is possible. What am I missing?

Reply:
The free pass is not intentional, will adjust the value. As for failedJobs > maxFailedJobRetries , the jobs are lexicographic in order even if we consider the timeStamp of LatestEntryTime the cursor will eventually move past these failed jobs and they will no longer be considered for processing since other jobs in the list will succeed except in the very very worst case scenario where all jobs keep failing. That is highly unlikely and would signal something else is wrong.

Reply:
made the updateFailedJob logic simpler